csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -2,246 +2,164 @@
+                ILocatorPrx newLocator = await GetLocatorAsync().ConfigureAwait(false);
+                if (locator != newLocator)
                 {
-                    while (!_locators.ContainsKey(instanceName) && _pending)
+                    var outgoingRequest = new OutgoingRequestFrame(
+                        newLocator, current.Operation, current.IsIdempotent, current.Context, incomingRequest.Payload);
+                    try
+                    {
+                        IncomingResponseFrame incomingResponse =
+                            await newLocator.InvokeAsync(outgoingRequest).ConfigureAwait(false);
+                        return new OutgoingResponseFrame(current.Encoding, incomingResponse.Payload);
+                    }
+                    catch (DispatchException)
+                    {
+                        throw;
+                    }
+                    catch (NoEndpointException)
+                    {
+                        throw new ObjectNotExistException(current);
+                    }
+                    catch (ObjectAdapterDeactivatedException)
+                    {
+                        throw new ObjectNotExistException(current);
+                    }
+                    catch (CommunicatorDestroyedException)
                     {
-                        Monitor.Wait(this, waitTime);
+                        throw new ObjectNotExistException(current);
+                    }
+                    catch (Exception ex)
+                    {
+                        lock (_mutex)
+                        {
+                            Debug.Assert((_locator == null && _completionSource != null) ||
+                                         (_locator != null && _completionSource == null));
+                            locator = newLocator;
+                            // If the current locator is equal to the one we use to send the request,
+                            // clear it and retry, this will trigger the lookup of a new locator.
+                            if (_locator == newLocator)
+                            {
+                                _locator = null;
+                            }
+                        }
+                        exception = ex;
                     }
                 }
-            }
-
-            //
-            // Return found locators
-            //
-            lock (this)
-            {
-                return new List<Ice.ILocatorPrx>(_locators.Values);
+                else
+                {
+                    // We got the same locator after a previous failure, throw the saved exception now.
+                    Debug.Assert(exception != null);
+                    throw exception;
+                }
             }
         }
 
-        public void
-        FoundLocator(ILocatorPrx? locator)
+        internal void FoundLocator(ILocatorPrx? locator)
         {
-            lock (this)
+            lock (_mutex)
```

Comment:
I would move the code from this method to the new FindLocatorAsync() method (see bellow). FoundLocator would just set the result task completion source to awake FindLocatorAsync and with the result an FindLocatorAsync.

Reply:
FoundLocator also deals with replies received after the completion source result was set, we cannot move that part to the FindLocatorAsync as it will be only awake for the first reply.